### PR TITLE
[core][distributed] add stateless process group

### PR DESCRIPTION
**Pull request by youkaichao**

### CODE REVIEW

**Comment by youkaichao** [comment]
merge for rlhf integration

File: vllm/distributed/device_communicators/pynccl.py [hunk 1]
```diff
@@ -33,13 +34,18 @@ def __init__(
         It is the caller's responsibility to make sure each communicator
         is bind to a unique device.
         """
-        assert dist.is_initialized()
-        assert dist.get_backend(group) != dist.Backend.NCCL, (
-            "PyNcclCommunicator should be attached to a non-NCCL group.")
+        if not isinstance(group, StatelessProcessGroup):
+            assert dist.is_initialized()
+            assert dist.get_backend(group) != dist.Backend.NCCL, (
+                "PyNcclCommunicator should be attached to a non-NCCL group.")
```
Comment (kwen2501) [inline vllm/distributed/device_communicators/pynccl.py:40]: typo?

**Review by kwen2501 — COMMENTED** [review]
LGTM.

General question:
(1) Do you want the new PG to be able to handle object collectives? If so, it seems its method surface would be similar to today PG's method surface. If that's the case, can we (PyTorch) re-work the `new_group` API so that it does not need calling `init_process_group` before it? IIUC, `init_process_group` is the main API that creates a global "state".

(2) If you don't need object collectives, would creating separate "Backend" objects work for you? I guess the "Backend" object would be similar to the PyncclCommunicator here.

**Comment by youkaichao** [comment]
> can we (PyTorch) re-work the new_group API so that it does not need calling init_process_group before it?

yes that would be great!

> would creating separate "Backend" objects work for you?

not sure how that works. i remember pytorch's nccl backend has some strange operations, like bucketing the transfer, and some logging wrapper for debugging. we don't want these.

File: vllm/distributed/device_communicators/pynccl.py (hunk 1 above)
Comment (youkaichao) [reply vllm/distributed/device_communicators/pynccl.py:40]: it's by design, we only use gloo group to initialize this nccl communicator. gloo for control-plane communication, while nccl for data plane.

**Pull request merged**